dotCMS/src/main/java/com/dotmarketing/services/PageServices.java

```diff
@@ -197,126 +197,142 @@ public static InputStream buildStream(IHTMLPage htmlPage, Identifier identifier,
-			sb.append("#set ($container").append(ident.getIdentifier() ).append( " = \"" ).append( folderPath ).append( ident.getIdentifier() ).append( "." ).append( Config.getStringProperty("VELOCITY_CONTAINER_EXTENSION") ).append( "\" )");
-
-			String sort = (c.getSortContentletsBy() == null) ? "tree_order" : c.getSortContentletsBy();
-
-			boolean dynamicContainer = UtilMethods.isSet(c.getLuceneQuery());
-
-			int langCounter = 0;
+        if(UtilMethods.isSet(containerList)) {
+			Iterator i = containerList.iterator();
+			while (i.hasNext()) {
+				Container ident = (Container) i.next();
+
+				Container c = null;
+				if (EDIT_MODE) {
+					c =
+						(Container) APILocator.getVersionableAPI()
+							.findWorkingVersion(ident.getIdentifier(), APILocator.getUserAPI().getSystemUser(), false);
+				} else {
+					c =
+						(Container) APILocator.getVersionableAPI()
+							.findLiveVersion(ident.getIdentifier(), APILocator.getUserAPI().getSystemUser(), false);
+				}
 
+				if (c == null)
+					continue;
+				//sets container to load the container file
+				sb.append("#set ($container").append(ident.getIdentifier()).append(" = \"").append(folderPath)
+					.append(ident.getIdentifier()).append(".")
+					.append(Config.getStringProperty("VELOCITY_CONTAINER_EXTENSION")).append("\" )");
+
+				String sort = (c.getSortContentletsBy() == null) ? "tree_order" : c.getSortContentletsBy();
+
+				boolean dynamicContainer = UtilMethods.isSet(c.getLuceneQuery());
+
+				int langCounter = 0;
+
+				List<Contentlet> contentlets = new ArrayList<Contentlet>();
+				List<Contentlet> contentletsFull = new ArrayList<Contentlet>();
+				if (!dynamicContainer) {
+					Identifier idenHtmlPage = APILocator.getIdentifierAPI().find(htmlPage);
+					Identifier idenContainer = APILocator.getIdentifierAPI().find(c);
+					//The container doesn't have categories
+					try {
+						contentlets =
+							conAPI.findPageContentlets(idenHtmlPage.getId(), idenContainer.getId(), sort, EDIT_MODE, -1,
+								APILocator.getUserAPI().getSystemUser(), false);
+						if (EDIT_MODE)
+							contentletsFull = contentlets;
+						else
+							contentletsFull =
+								conAPI.findPageContentlets(idenHtmlPage.getId(), idenContainer.getId(), sort, true, -1,
+									APILocator.getUserAPI().getSystemUser(), false);
+					} catch (Exception e) {
+						Logger.error(PageServices.class, "Unable to retrive contentlets on page", e);
+					}
+					Logger.debug(PageServices.class,
+						"HTMLPage= " + htmlPage.getInode() + " Container=" + c.getInode() + " Language=-1 Contentlets="
+							+ contentlets.size());
+				}
+				//this is to filter the contentlets list removing the repited identifiers
+				if (contentlets.size() > 0) {
+					Set<String> contentletIdentList = new HashSet<String>();
+					List<Contentlet> contentletsFilter = new ArrayList<Contentlet>();
+					for (Contentlet cont : contentlets) {
+						if (!contentletIdentList.contains(cont.getIdentifier())) {
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this code to not nest more than 4 if/for/while/switch/try statements. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS134)